sdk/test-utils/perfstress/src/testProxyHttpClient.ts

```diff
@@ -0,0 +1,289 @@
+    }
+    this._recordingId = id;
+    this.stateManager.setState("started-recording");
+  }
+
+  async stopRecording(): Promise<void> {
+    this.stateManager.validateState("stopping-recording");
+    const stopUri = this._uri + paths.record + paths.stop;
+    const req = this._createRecordingRequest(stopUri);
+    req.headers.set("x-recording-id", this._recordingId!);
+    await this._httpClient.sendRequest(req);
+    this.stateManager.setState("stopped-recording");
+  }
+
+  async startPlayback(): Promise<void> {
+    this.stateManager.validateState("starting-playback");
+    const startUri = this._uri + paths.playback + paths.start;
+    const req = this._createRecordingRequest(startUri);
+    req.headers.set("x-recording-id", this._recordingId!);
+    const rsp = await this._httpClient.sendRequest(req);
+    if (rsp.status !== 200) {
+      throw new Error("Start request failed.");
+    }
+    const id = rsp.headers.get("x-recording-id");
+    if (!id) {
+      throw new Error("No recording ID returned.");
+    }
+    this._recordingId = id;
+    this.stateManager.setState("started-playback");
+  }
+
+  async stopPlayback(): Promise<void> {
+    this.stateManager.validateState("stopping-playback");
+    const stopUri = this._uri + paths.playback + paths.stop;
+    const req = this._createRecordingRequest(stopUri);
+    req.headers.set("x-recording-id", this._recordingId!);
+    req.headers.set("x-purge-inmemory-recording", "true");
+    await this._httpClient.sendRequest(req);
+
+    this._mode = "live";
+    this._recordingId = undefined;
+    this.stateManager.setState("stopped-playback");
+  }
+
+  private _createRecordingRequest(uri: string) {
+    const req = new WebResource(uri, "POST");
+    if (this._recordingId !== undefined) {
+      req.headers.set("x-recording-id", this._recordingId);
+    }
+    return req;
+  }
+}
+
+export class TestProxyHttpClient {
+  private _uri: string;
+  private _httpClient: HttpClient;
+  private _recordingId?: string;
+  public _mode!: string;
+  private stateManager: RecordingStateManager = new RecordingStateManager();
+
+  constructor(uri: string) {
+    this._uri = uri;
+    this._httpClient = createDefaultHttpClient();
+  }
+
+  async sendRequest(request: PipelineRequest): Promise<PipelineRequest> {
+    if (this._recordingId && (this._mode === "record" || this._mode === "playback")) {
+      request.headers.set("x-recording-id", this._recordingId);
+      request.headers.set("x-recording-mode", this._mode);
+      request.headers.set("x-recording-remove", "false");
+
+      const upstreamUrl = new URL(request.url);
+      const redirectedUrl = new URL(request.url);
+
+      const providedUrl = new URL(this._uri);
+      redirectedUrl.host = providedUrl.host;
+      redirectedUrl.port = providedUrl.port;
+      redirectedUrl.protocol = providedUrl.protocol;
+      upstreamUrl.pathname = "";
+      request.headers.set("x-recording-upstream-base-uri", upstreamUrl.toString());
```

Comment:
If JS is similar to .NET, you need to ensure this header is only set once, since this codepath may execute multiple times on retries:

https://github.com/Azure/azure-sdk-for-net/pull/22621/files#diff-b2fb8a7a06a2f30913d5933b577bf9e6763764c19af5e841ec450c2b360bfe2aR47

Reply:
Done